### PR TITLE
[ticket/8672] Add class for retrieving imagesize without download

### PR DESCRIPTION
**Pull request by marc1706**
getimagesize() always downloads the complete file before checking
the actual image dimensions. This class will be able to do the same
without having to download possibly large files.

Ticket: https://tracker.phpbb.com/browse/PHPBB3-8672

Replaces https://github.com/phpbb/phpbb/pull/3382

### CODE REVIEW

**Comment by bantu** [comment]
I have to ask: What existing solutions for this problem exist? How does this library compare to existing libraries?

**Comment by marc1706** [comment]
There don't seem to be many libraries that target this specific issue with getimagesize().
Libraries that I was able to find are:

* https://github.com/MatejB/rim
* https://github.com/tommoor/fastimage

Both of these can't be installed via composer. In addition to that, they only support a small amount of image types (gif, jpeg, png, bmp).

Also, the rim library is licensed under the LGPLv3 license which is not compatible with the GNU GPL v2 and it uses CURL to retrieve files which we probably don't want to require.

File: build/build.xml [hunk 1]
```diff
@@ -434,6 +434,15 @@
 		<delete file="${dir}/vendor/symfony/twig-bridge/Symfony/Bridge/Twig/CHANGELOG.md" />
 		<delete file="${dir}/vendor/symfony/twig-bridge/Symfony/Bridge/Twig/README.md" />
 		<delete file="${dir}/vendor/symfony/twig-bridge/Symfony/Bridge/Twig/phpunit.xml.dist" />
+
+        <delete dir="${dir}/vendor/marc1706/fast-image-size/tests" />
```
Comment (dhruvgoel92) [inline build/build.xml]: indentation is off a bit here.

**Comment by Nicofuma** [comment]
+1

File: build/build.xml (hunk 1 above)
Comment (Nicofuma) [inline build/build.xml]: actually he used spaces instead of tabs. ;)

File: build/build.xml [hunk 2]
```diff
@@ -434,6 +434,15 @@
...
+        <delete dir="${dir}/vendor/marc1706/fast-image-size/travis" />
+        <delete dir="${dir}/vendor/marc1706/fast-image-size/.git" />
+        <delete file="${dir}/vendor/marc1706/fast-image-size/.gitignore" />
+        <delete file="${dir}/vendor/marc1706/fast-image-size/.travis.yml" />
+        <delete file="${dir}/vendor/marc1706/fast-image-size/composer.phar" />
+        <delete file="${dir}/vendor/marc1706/fast-image-size/phpunit.xml.dist" />
+        <delete file="${dir}/vendor/marc1706/fast-image-size/README.md" />
```
Comment (Nicofuma) [inline build/build.xml]: tabs please

**Comment by marc1706** [comment]
Got rid of the needed additions to clean-vendor-dir :)

File: phpBB/composer.json [hunk 3]
```diff
@@ -29,6 +29,7 @@
 		"lusitanian/oauth": "0.2.*",
 		"patchwork/utf8": "1.1.*",
 		"s9e/text-formatter": "dev-release/php5.3",
+		"marc1706/fast-image-size": "1.0.*",
```
Comment (Nicofuma) [inline phpBB/composer.json]: alphabetical order please :)

File: phpBB/includes/functions_upload.php [hunk 4]
```diff
@@ -387,28 +387,28 @@ function move_file($destination, $overwrite = false, $skip_image_check = false,
...
-					if (!isset($types[$this->image_info[2]]))
+					if (!isset($types[$this->image_info['type']]))
 					{
-						$this->error[] = sprintf($user->lang['IMAGE_FILETYPE_INVALID'], $this->image_info[2], $this->mimetype);
+						$this->error[] = sprintf($user->lang['IMAGE_FILETYPE_INVALID'], $this->image_info['type'], $this->mimetype);
 					}
 					else
 					{
-						$this->error[] = sprintf($user->lang['IMAGE_FILETYPE_MISMATCH'], $types[$this->image_info[2]][0], $this->extension);
+						$this->error[] = sprintf($user->lang['IMAGE_FILETYPE_MISMATCH'], $types[$this->image_info['type']][0], $this->extension);
```
Comment (Nicofuma) [inline phpBB/includes/functions_upload.php]: $user->lang() by the way :)

**Comment by marc1706** [comment]
@Nicofuma modified as requested and rebased onto current master

**Comment by marc1706** [comment]
@Nicofuma rebased again

**Comment by Nicofuma** [comment]
If I try to upload a too big image I don't have any error and an image is actually uploaded and stored: a 70 bytes image

**Comment by marc1706** [comment]
Please define what you mean by too big? File size or image dimensions?

**Comment by Nicofuma** [comment]
image dimensions

**Comment by marc1706** [comment]
```
<Marc> Nicofuma: so you restricted the image dimensions in the acp and uploaded an image to a post afterwards, correct?
<Marc> is it possible that plupload resized the image for you? :D
<Nicofuma> No idea
<Marc> was it a valid image?
<Nicofuma> Yes
<Marc> just a tiny version of it? :D
<Marc> yeah, then plupload resized the image for you :)
<Nicofuma> I set it to 1x1 so ^^^
<Marc> hehe
<Marc> yeah, plupload :D
<Marc> yup, just checked
<Marc> it's creating a 1x1 picture :D
```

**Pull request merged**